libcloud/compute/drivers/packet.py

```diff
@@ -0,0 +1,254 @@
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Packet Driver
+"""
+
+from libcloud.utils.py3 import httplib
+
+from libcloud.common.base import ConnectionUserAndKey
+from libcloud.common.base import JsonResponse
+from libcloud.compute.types import Provider, NodeState, InvalidCredsError
+from libcloud.compute.base import NodeDriver, Node
+from libcloud.compute.base import NodeImage, NodeSize, NodeLocation
+
+PACKET_ENDPOINT = "api.packet.net"
+
+
+class PacketResponse(JsonResponse):
+    valid_response_codes = [httplib.OK, httplib.ACCEPTED, httplib.CREATED,
+                            httplib.NO_CONTENT]
+
+    def parse_error(self):
+        if self.status == httplib.UNAUTHORIZED:
+            body = self.parse_body()
+            raise InvalidCredsError(body['message'])
+        else:
+            body = self.parse_body()
+            if 'message' in body:
+                error = '%s (code: %s)' % (body['message'], self.status)
+            else:
+                error = body
+            return error
+
+    def success(self):
+        return self.status in self.valid_response_codes
+
+
+class SSHKey(object):
+    def __init__(self, id, label, key):
+        self.id = id
+        self.label = label
+        self.key = key
+
+    def __repr__(self):
+        return (('<SSHKey: id=%s, label=%s, key=%s>') %
+                (self.id, self.label, self.key))
+
+
+class PacketConnection(ConnectionUserAndKey):
+    """
+    Connection class for the Packet driver.
+    """
+
+    host = PACKET_ENDPOINT
+    responseCls = PacketResponse
+
+    def add_default_headers(self, headers):
+        """
+        Add headers that are necessary for every request
+        """
+        headers['Content-Type'] = 'application/json'
+        headers['X-Auth-Token'] = self.user_id
+        headers['X-Consumer-Token'] = \
```

Comment:
Ah, I see, that makes sense. Thanks for the clarification.

P.S. We also try to be a good citizen and indicate that the request is made by Libcloud by sending a right user-agent header (e.g. `User-Agent: libcloud/0.17.1-dev (Rackspace Cloud (Next Gen))` :)).